pkg/operator/ceph/cluster/dependents.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright 2020 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	"github.com/rook/rook/pkg/clusterd"
+	"github.com/rook/rook/pkg/util"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	// must use plural kinds
+	cephClusterDependentPluralKinds []string = []string{
+		"CephBlockPools",
+		"CephRBDMirrors",
+		"CephFilesystems",
+		"CephFilesystemMirrors",
+		"CephObjectStores",
+		"CephObjectStoreUsers",
+		"CephObjectZones",
+		"CephObjectZoneGroups",
+		"CephObjectRealms",
+		"CephNFSes",
+		"CephClients",
+	}
+)
+
+// CephClusterDependents returns a DependentList of dependents of a CephCluster in the namespace.
+func CephClusterDependents(c *clusterd.Context, namespace string) (*DependentList, error) {
+	ctx := context.TODO()
+
+	dependents := NewDependentList()
+	errs := []error{}
+
+	for _, pluralKind := range cephClusterDependentPluralKinds {
+		resource := pluralKindToResource(pluralKind)
+		gvr := cephv1.SchemeGroupVersion.WithResource(resource)
+		list, err := c.DynamicClientset.Resource(gvr).Namespace(namespace).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			errs = append(errs, errors.Wrapf(err, "failed to list %s", pluralKind))
```

Comment:
Why do we need to aggregate errors? If there is any error it seems like the caller can't trust the results.

Reply:
It's true, but fully expressing the errors to the user is good for debugging rather than short-circuiting at the first error which could hide multiple errors.

Reply:
In this case it seems like only api errors would be reported, in which case the same error may be returned multiple times, which would be more overwhelming than helpful. I'd still vote to return on first error. If it's intermittent, the next reconcile will anyway succeed and it will just be a transient log message.

Reply:
On more general thought about where my comment is coming from... Simplicity always comes first as a design principle. Complexity should only be added where complexity is needed.

Reply:
I am coming from a user perspective of things. It is really annoying when one error obfuscates other errors when all errors could be reported at once and is sometimes valuable for understanding the underlying cause. Additionally, it is best for the user experience to do our best to find all of the blocking resources that we can in order to report them so that users get as much of the correct information as we can give.

Reply:
When there are multiple resources that contribute to blocking the CR, agreed that we should include them all in the report so they can all be addressed. And you're doing that in another place. 

In this instance, it's just likely reporting a transient api error. A retry at the next reconcile will probably resolve itself anyway. But if all ~10 api calls fail, we will likely get the same error 10 times and it will just fill the log with duplicate messages, right? Or do you have an example of multiple actionable errors that could come from this?

Reply:
I think that if all 10 API calls fail, that could be helpful information compared to just knowing that one failed. That might imply an API server failure or network partition. Or if we can see that one (or a few) repeatedly fails when others do not, that might suggest that the user has malformed or forgotten to register those CRDs/types. And if we return all of the failures, we can know exactly which CRDs/types aren't registered, which might help us determine that the issue is missing an upgrade doc step versus a one-off issue.

Reply:
It still seems like a stretch to need the aggregate errors in this case, but the complexity of continuing this conversation is not worth the request for simplicity. :)